<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Estimating Microenvironment from Spatial Data • NICHES</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Estimating Microenvironment from Spatial Data">
<meta property="og:description" content="NICHES">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">NICHES</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01%20NICHES%20Spatial.html">Estimating Microenvironment from Spatial Data</a>
    </li>
    <li>
      <a href="../articles/02%20NICHES%20Single.html">Converting Single-Cell Atlases to Single-Cell Signaling Atlases</a>
    </li>
    <li>
      <a href="../articles/03%20Rat%20Alveolus.html">Cellular Relationships in the Rat Alveolus</a>
    </li>
    <li>
      <a href="../articles/04%20Differential%20CellToCell.html">Differential CellToCell Signaling Across Conditions</a>
    </li>
    <li>
      <a href="../articles/05%20Differential%20SystemToCell.html">Differential SystemToCell Signaling Across Conditions</a>
    </li>
    <li>
      <a href="../articles/06%20Pseudotime.html">Pseudotemporal Ordering with NICHES</a>
    </li>
    <li>
      <a href="../articles/07%20Spatiotemporal%20NICHES.html">Spatio-Temporal NICHES</a>
    </li>
    <li>
      <a href="../articles/08%20Interperobility%20-%20NICHES.html">08 Interperobility - NICHES</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Estimating Microenvironment from Spatial Data</h1>
                        <h4 class="author">Junchen Yang &amp; Micha Sam Brickman Raredon</h4>
            
            <h4 class="date">2022-07-05</h4>
      
      
      <div class="hidden name"><code>01 NICHES Spatial.Rmd</code></div>

    </div>

    
    
<p>NICHES is a toolset which transforms single-cell atlases into single-cell-signaling atlases. It is engineered to be computationally efficient and very easy to run. It interfaces directly with Seurat from Satija Lab. The cell-signaling outputs from NICHES may be analyzed with any single-cell toolset, including Seurat, Scanpy, Monocle, or others.</p>
<p>Here, we show how NICHES may be used to estimate individual cellular microenvironment from spatial transcriptomic data.</p>
<p>First, let’s load dependencies.</p>
<div id="load-dependencies" class="section level2">
<h2 class="hasAnchor">
<a href="#load-dependencies" class="anchor"></a>Load Dependencies</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat">SeuratData</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SeuratWrappers</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://msraredon.github.io/NICHES/">NICHES</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sjmgarnier/viridis">viridis</a></span><span class="op">)</span></code></pre></div>
<p>Next, we load the data, perform basic pre-processing, and cluster the data so that we can visualize patterns of interest. For this vignette we will use basic Seurat clustering annotations to avoid the work of labeling celltypes, which are not necessary for this demonstration.</p>
</div>
<div id="load-data-normalize-visualize" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data-normalize-visualize" class="anchor"></a>Load Data, Normalize, Visualize</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html">InstallData</a></span><span class="op">(</span><span class="st">"stxBrain"</span><span class="op">)</span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html">LoadData</a></span><span class="op">(</span><span class="st">"stxBrain"</span>, type <span class="op">=</span> <span class="st">"anterior1"</span><span class="op">)</span>
<span class="co"># Normalization </span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">brain</span>, assay <span class="op">=</span> <span class="st">"Spatial"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">brain</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Hpca"</span>, <span class="st">"Ttr"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="01%20NICHES%20Spatial_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Dimensional reduction with all cells</span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">brain</span>, assay <span class="op">=</span> <span class="st">"SCT"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">brain</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">brain</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">brain</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">brain</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>,group.by <span class="op">=</span> <span class="st">'seurat_clusters'</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html">SpatialDimPlot</a></span><span class="op">(</span><span class="va">brain</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,group.by <span class="op">=</span> <span class="st">'seurat_clusters'</span>, label.size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></code></pre></div>
<p><img src="01%20NICHES%20Spatial_files/figure-html/unnamed-chunk-2-2.png" width="700"></p>
<p>These numeric annotations are satisfactory for the demonstration of NICHES. Any and all metadata can be carried over when NICHES is run, allowing multiple levels of clustering and sub-clustering to be leveraged downstream after micorenvironment calculation.</p>
<p>Next, we will format the spatial coordinate metadata so that every cell has an explicitly labeled x and y coordinate.</p>
</div>
<div id="format-spatial-coordinates-and-normalize" class="section level2">
<h2 class="hasAnchor">
<a href="#format-spatial-coordinates-and-normalize" class="anchor"></a>Format Spatial Coordinates and Normalize</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">brain</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">brain</span><span class="op">@</span><span class="va">images</span><span class="op">$</span><span class="va">anterior1</span><span class="op">@</span><span class="va">coordinates</span><span class="op">$</span><span class="va">row</span>
<span class="va">brain</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">brain</span><span class="op">@</span><span class="va">images</span><span class="op">$</span><span class="va">anterior1</span><span class="op">@</span><span class="va">coordinates</span><span class="op">$</span><span class="va">col</span>

<span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">brain</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Spatial"</span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">brain</span><span class="op">)</span></code></pre></div>
<p>NICHES can be run on imputed or non-imputed data. Here, we will use imputed data.</p>
</div>
<div id="impute-and-run-niches" class="section level2">
<h2 class="hasAnchor">
<a href="#impute-and-run-niches" class="anchor"></a>Impute and Run NICHES</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu">SeuratWrappers</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratWrappers/man/RunALRA.html">RunALRA</a></span><span class="op">(</span><span class="va">brain</span><span class="op">)</span>

<span class="va">NICHES_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunNICHES.html">RunNICHES</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">brain</span>,
                           LR.database <span class="op">=</span> <span class="st">"fantom5"</span>,
                           species <span class="op">=</span> <span class="st">"mouse"</span>,
                           assay <span class="op">=</span> <span class="st">"alra"</span>,
                           position.x <span class="op">=</span> <span class="st">'x'</span>,
                           position.y <span class="op">=</span> <span class="st">'y'</span>,
                           k <span class="op">=</span> <span class="fl">4</span>, 
                           cell_types <span class="op">=</span> <span class="st">"seurat_clusters"</span>,
                           min.cells.per.ident <span class="op">=</span> <span class="fl">0</span>,
                           min.cells.per.gene <span class="op">=</span> <span class="cn">NULL</span>,
                           meta.data.to.map <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'orig.ident'</span>,<span class="st">'seurat_clusters'</span><span class="op">)</span>,
                           CellToCell <span class="op">=</span> <span class="cn">F</span>,CellToSystem <span class="op">=</span> <span class="cn">F</span>,SystemToCell <span class="op">=</span> <span class="cn">F</span>,
                           CellToCellSpatial <span class="op">=</span> <span class="cn">F</span>,CellToNeighborhood <span class="op">=</span> <span class="cn">F</span>,NeighborhoodToCell <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>NICHES outputs a list of objects. Each object contains a certain style of cell-system signaling atlas. Above, we have only calculated a single one of interest, namely, individual cellular microenvironment. We next isolate this output and embed using UMAP to visualize the microenvironemnt of each cell.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">niche</span> <span class="op">&lt;-</span> <span class="va">NICHES_output</span><span class="op">[[</span><span class="st">'NeighborhoodToCell'</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">niche</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">niche</span><span class="op">[[</span><span class="st">'ReceivingType'</span><span class="op">]</span><span class="op">]</span>

<span class="co"># Scale and visualize</span>
<span class="va">niche</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">niche</span><span class="op">)</span>
<span class="va">niche</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">niche</span>,selection.method <span class="op">=</span> <span class="st">"disp"</span><span class="op">)</span>
<span class="va">niche</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">niche</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">niche</span>,ndims <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></code></pre></div>
<p><img src="01%20NICHES%20Spatial_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">niche</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">niche</span>,dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">niche</span>,reduction <span class="op">=</span> <span class="st">'umap'</span>,pt.size <span class="op">=</span> <span class="fl">0.5</span>,shuffle <span class="op">=</span> <span class="cn">T</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Cellular Microenvironment'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="01%20NICHES%20Spatial_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<p>We can already see, from this plot, some notable overlap between the microenvironments of celltypes 1 &amp; 7 and celltypes 6 &amp; 3. Let’s explore this more deeply by finding signaling mechanisms specific to each celltype niche, plotting some of the results in heatmap form:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Find markers</span>
<span class="va">mark</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">niche</span>,min.pct <span class="op">=</span> <span class="fl">0.25</span>,only.pos <span class="op">=</span> <span class="cn">T</span>,test.use <span class="op">=</span> <span class="st">"roc"</span><span class="op">)</span>
<span class="va">GOI_niche</span> <span class="op">&lt;-</span> <span class="va">mark</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span><span class="fl">5</span>,<span class="va">myAUC</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="va">niche</span>,features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">GOI_niche</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>,<span class="st">"white"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="01%20NICHES%20Spatial_files/figure-html/unnamed-chunk-6-1.png" width="960"></p>
<p>This confirms that celltypes 1 &amp; 7 and 6 &amp; 3 do indeed have some shared character.</p>
<p>We can further confirm that identified celltype specific signaling mechanisms are indeed specific to tissue regions in which those cells are found, by plotting matched ligand and receptor pairs:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Check that these make sense and print little plots</span>
<span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">brain</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'alra'</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">brain</span>, crop <span class="op">=</span> <span class="cn">TRUE</span>, features <span class="op">=</span> <span class="st">"Fgf1"</span>,slot <span class="op">=</span> <span class="st">"data"</span>,min.cutoff <span class="op">=</span>  <span class="st">'q1'</span>,
                         max.cutoff <span class="op">=</span> <span class="st">'q99'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Ligand"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">brain</span>, crop <span class="op">=</span> <span class="cn">TRUE</span>, features <span class="op">=</span> <span class="st">"Fgfr2"</span>,slot <span class="op">=</span> <span class="st">"data"</span>,min.cutoff <span class="op">=</span>  <span class="st">'q1'</span>,
                         max.cutoff <span class="op">=</span> <span class="st">'q99'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Receptor"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="va">p1</span>,<span class="va">p2</span><span class="op">)</span></code></pre></div>
<p><img src="01%20NICHES%20Spatial_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Further, and perhaps more usefully, we can map over the output from NICHES onto the original spatial object as follows:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Add Niches output as an assay</span>
<span class="va">niches.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span>  <span class="va">niche</span><span class="op">[[</span><span class="st">'NeighborhoodToCell'</span><span class="op">]</span><span class="op">]</span>, slot <span class="op">=</span> <span class="st">'data'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">niches.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">niche</span><span class="op">[[</span><span class="st">'ReceivingCell'</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ReceivingCell</span>
<span class="va">brain</span><span class="op">[[</span><span class="st">"NeighborhoodToCell"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateAssayObject.html">CreateAssayObject</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">niches.data</span> <span class="op">)</span>
<span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">brain</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"NeighborhoodToCell"</span>
<span class="va">brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">brain</span><span class="op">)</span></code></pre></div>
<p>Which allows direct visualization of niche interactions of interest in a spatial context:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot celltype specific niche signaling</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">brain</span>,
                   features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Bmp2—Bmpr2'</span>,<span class="st">'Efna1—Ephb6'</span>,<span class="st">'Fgf1—Fgfr2'</span><span class="op">)</span>,
                   slot <span class="op">=</span> <span class="st">'scale.data'</span><span class="op">)</span></code></pre></div>
<p><img src="01%20NICHES%20Spatial_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Micha Sam Brickman Raredon, Junchen Yang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
