<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interoperability • NICHES</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Interoperability">
<meta property="og:description" content="NICHES">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">NICHES</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01%20NICHES%20Spatial.html">Estimating Microenvironment from Spatial Data</a>
    </li>
    <li>
      <a href="../articles/02%20NICHES%20Single.html">Converting Single-Cell Atlases to Single-Cell Signaling Atlases</a>
    </li>
    <li>
      <a href="../articles/03%20Rat%20Alveolus.html">Cellular Relationships in the Rat Alveolus</a>
    </li>
    <li>
      <a href="../articles/04%20Differential%20CellToCell.html">Differential CellToCell Signaling Across Conditions</a>
    </li>
    <li>
      <a href="../articles/05%20Differential%20SystemToCell.html">Differential SystemToCell Signaling Across Conditions</a>
    </li>
    <li>
      <a href="../articles/06%20Pseudotime.html">Pseudotemporal Ordering with NICHES</a>
    </li>
    <li>
      <a href="../articles/07%20Spatiotemporal%20NICHES.html">Spatio-Temporal NICHES</a>
    </li>
    <li>
      <a href="../articles/08%20Interoperability%20-%20NICHES.html">Interoperability</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Interoperability</h1>
                        <h4 class="author">Junchen Yang</h4>
            
            <h4 class="date">2022-07-10</h4>
      
      
      <div class="hidden name"><code>08 Interoperability - NICHES.Rmd</code></div>

    </div>

    
    
<p>In the previous vignettes, NICHES was mainly demonstrated with Seurat framework. Here in this vignette, we demonstrate the interoperability of NICHES with other popular scRNA-seq frameworks, namely, Scanpy, Monocle3, and scater.</p>
<p>First, let’s start with Scanpy and use the pbmc3k dataset as a demonstration. Scanpy is python-based and uses AnnData as the primary data strcture. We will demonstrate how to extract the data from Scanpy to run NICHES, convert the NICHES output to AnnData, and use functions from Scanpy to perform downstream analysis.</p>
<div id="interperobility-with-scanpy" class="section level1">
<h1 class="hasAnchor">
<a href="#interperobility-with-scanpy" class="anchor"></a>Interperobility with Scanpy</h1>
<div id="load-required-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#load-required-packages" class="anchor"></a>Load Required Packages</h2>
<p>Here besides NICHES, we will also load ALRA for imputation and reticulate for running python.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://msraredon.github.io/NICHES/">NICHES</a></span><span class="op">)</span>

<span class="co"># Load ALRA</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/source_url.html">source_url</a></span><span class="op">(</span><span class="st">'https://raw.githubusercontent.com/KlugerLab/ALRA/master/alra.R'</span><span class="op">)</span>

<span class="co"># Set up reticulate and the python env</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/">reticulate</a></span><span class="op">)</span>
<span class="co"># change to your corresponding python path</span>
<span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html">Sys.setenv</a></span><span class="op">(</span>RETICULATE_PYTHON <span class="op">=</span> <span class="st">"/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10"</span><span class="op">)</span></code></pre></div>
<p>We will also need a few python packages, in particular, scanpy.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="im">import</span> numpy <span class="im">as</span> np</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="im">import</span> scanpy <span class="im">as</span> sc</a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb3-1" data-line-number="1">sc.set_figure_params(dpi<span class="op">=</span><span class="dv">300</span>,figsize<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">16</span>))</a></code></pre></div>
</div>
<div id="load-the-data-and-preprocess" class="section level2">
<h2 class="hasAnchor">
<a href="#load-the-data-and-preprocess" class="anchor"></a>Load the data and preprocess</h2>
<p>Now let’s load the pbmc3k data from scanpy directly. Here we load both the raw and processed pbmc3k. The former will be used to run NICHES after a few processing steps and we will use the latter to add cell annotations.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># load the raw pbmc3k and the processed pbmc3k</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">adata <span class="op">=</span> sc.datasets.pbmc3k()</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">adata2 <span class="op">=</span> sc.datasets.pbmc3k_processed()</a></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Filtering</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">sc.pp.filter_cells(adata, min_genes<span class="op">=</span><span class="dv">200</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">sc.pp.filter_genes(adata, min_cells<span class="op">=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">adata.var[<span class="st">'mt'</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">'MT-'</span>)</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">'mt'</span>], percent_top<span class="op">=</span><span class="va">None</span>, log1p<span class="op">=</span><span class="va">False</span>, inplace<span class="op">=</span><span class="va">True</span>)</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">adata <span class="op">=</span> adata[adata.obs.n_genes_by_counts <span class="op">&lt;</span> <span class="dv">2500</span>, :]</a>
<a class="sourceLine" id="cb5-7" data-line-number="7">adata <span class="op">=</span> adata[adata.obs.pct_counts_mt <span class="op">&lt;</span> <span class="dv">5</span>, :]</a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co"># Log1p normalization</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9">sc.pp.normalize_total(adata, target_sum<span class="op">=</span><span class="fl">1e4</span>)</a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt; /Users/mbr29/Library/Python/3.10/lib/python/site-packages/scanpy/preprocessing/_normalization.py:170: UserWarning: Received a view of an AnnData. Making a copy.</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt;   view_to_actual(adata)</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12">sc.pp.log1p(adata)</a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co"># Assign labels</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14">adata.obs[<span class="st">'louvain'</span>] <span class="op">=</span> adata2.obs[<span class="st">'louvain'</span>]</a></code></pre></div>
</div>
<div id="visualize-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-the-data" class="anchor"></a>Visualize the data</h2>
<p>We can visualize the processed data on UMAP using Scanpy.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="cf">with</span> plt.rc_context({<span class="st">"figure.figsize"</span>: (<span class="dv">8</span>, <span class="dv">8</span>), <span class="st">"figure.dpi"</span>: (<span class="dv">300</span>)}):</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">    sc.pl.umap(adata2, color<span class="op">=</span><span class="st">'louvain'</span>,legend_loc <span class="op">=</span><span class="st">"on data"</span>)</a></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-4-1.png" width="768"></p>
</div>
<div id="subset-to-only-b-cellscd4-t-cellscd14-monocytes" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-to-only-b-cellscd4-t-cellscd14-monocytes" class="anchor"></a>Subset to only ‘B cells’,‘CD4 T cells’,‘CD14+ Monocytes’</h2>
<p>Here for demonstration, For let’s focus in on signaling within a simple 3-cell system:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb7-1" data-line-number="1">adata <span class="op">=</span> adata[adata.obs[<span class="st">'louvain'</span>].isin([<span class="st">'B cells'</span>,<span class="st">'CD4 T cells'</span>,<span class="st">'CD14+ Monocytes'</span>]),:]</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">cell_names <span class="op">=</span> np.array(adata.obs_names)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">gene_names <span class="op">=</span> np.array(adata.var_names)</a></code></pre></div>
</div>
<div id="extract-the-meta-data-and-impute-the-data-to-run-niches" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-the-meta-data-and-impute-the-data-to-run-niches" class="anchor"></a>Extract the meta data and impute the data to run NICHES</h2>
<p>Here we will extract the cell types metadata from anndata and impute the data to run NICHES. In particular, data.imputed will the data matrix to run NICHES.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta.data.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"louvain"</span> <span class="op">=</span> <span class="va">py</span><span class="op">$</span><span class="va">adata</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">louvain</span>,row.names <span class="op">=</span> <span class="va">py</span><span class="op">$</span><span class="va">cell_names</span><span class="op">)</span>

<span class="co"># alra imputation</span>
<span class="va">imputed.list</span> <span class="op">&lt;-</span> <span class="fu">alra</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">py</span><span class="op">$</span><span class="va">adata</span><span class="op">$</span><span class="va">X</span><span class="op">$</span><span class="fu">toarray</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Read matrix with 1966 cells and 13714 genes</span>
<span class="co">#&gt; Chose k=13</span>
<span class="co">#&gt; Getting nonzeros</span>
<span class="co">#&gt; Randomized SVD</span>
<span class="co">#&gt; Find the 0.001000 quantile of each gene</span>
<span class="co">#&gt; Sweep</span>
<span class="co">#&gt; Scaling all except for 418 columns</span>
<span class="co">#&gt; 0.00% of the values became negative in the scaling process and were set to zero</span>
<span class="co">#&gt; The matrix went from 5.89% nonzero to 46.50% nonzero</span>
<span class="va">data.imputed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">imputed.list</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data.imputed</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">py</span><span class="op">$</span><span class="va">cell_names</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data.imputed</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">py</span><span class="op">$</span><span class="va">gene_names</span></code></pre></div>
</div>
<div id="run-niches" class="section level2">
<h2 class="hasAnchor">
<a href="#run-niches" class="anchor"></a>Run NICHES</h2>
<p>Here we will take the previously imputed data matrix as input to run NICHES and specify output_format to be “raw” for easier conversion to AnnData.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">niches_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunNICHES.html">RunNICHES</a></span><span class="op">(</span><span class="va">data.imputed</span>,
                 meta.data.df <span class="op">=</span> <span class="va">meta.data.df</span>,
                 cell_types <span class="op">=</span> <span class="st">"louvain"</span>,
                 species <span class="op">=</span> <span class="st">'human'</span>,
                 LR.database <span class="op">=</span> <span class="st">'fantom5'</span>,
                 CellToCell <span class="op">=</span> <span class="cn">T</span>,output_format <span class="op">=</span> <span class="st">"raw"</span>
                 <span class="op">)</span>
<span class="va">lr_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">niches_list</span><span class="op">$</span><span class="va">CellToCell</span><span class="op">$</span><span class="va">CellToCellMatrix</span><span class="op">)</span></code></pre></div>
</div>
<div id="convert-the-output-to-anndata" class="section level2">
<h2 class="hasAnchor">
<a href="#convert-the-output-to-anndata" class="anchor"></a>Convert the output to AnnData</h2>
<p>We will extract the output CellToCellMatrix from NICHES and convert it to AnnData.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb10-1" data-line-number="1">adata_niches <span class="op">=</span> sc.AnnData(X <span class="op">=</span> r.niches_list[<span class="st">'CellToCell'</span>][<span class="st">'CellToCellMatrix'</span>].T)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co">#&gt; &lt;string&gt;:1: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">adata_niches.obs <span class="op">=</span> r.niches_list[<span class="st">'CellToCell'</span>][<span class="st">'metadata'</span>]</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">adata_niches.var_names <span class="op">=</span> r.lr_names</a></code></pre></div>
</div>
<div id="process-the-output-from-niches-using-scanpy" class="section level2">
<h2 class="hasAnchor">
<a href="#process-the-output-from-niches-using-scanpy" class="anchor"></a>Process the output from NICHES using Scanpy</h2>
<p>We will use the Scanpy functions to perform downstream analysis. Here we will scale the data, do PCA and UMAP.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># data scaling, pca, find neighbors, and umap</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">sc.pp.scale(adata_niches)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">sc.tl.pca(adata_niches)</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">sc.pp.neighbors(adata_niches)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">sc.tl.umap(adata_niches)</a></code></pre></div>
</div>
<div id="visualize-celltocell-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-celltocell-matrix" class="anchor"></a>Visualize CellToCell matrix</h2>
<p>We can use the visualization function from Scanpy to visualize the data in UMAP.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="cf">with</span> plt.rc_context({<span class="st">"figure.figsize"</span>: (<span class="dv">8</span>, <span class="dv">8</span>), <span class="st">"figure.dpi"</span>: (<span class="dv">300</span>)}):</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">    sc.pl.umap(adata_niches,color<span class="op">=</span><span class="st">"VectorType"</span>,legend_loc <span class="op">=</span><span class="st">"on data"</span>,legend_fontsize<span class="op">=</span><span class="dv">7</span>)</a></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
</div>
<div id="differential-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-analysis" class="anchor"></a>Differential analysis</h2>
<p>We can also use the function from Scanpy to perform differential analysis between different VectorTypes. Here we will rank the genes using a logistic regression classifier with L1 regularization and visualize the top markers.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb13-1" data-line-number="1">sc.tl.rank_genes_groups(adata_niches, <span class="st">'VectorType'</span>, method<span class="op">=</span><span class="st">'logreg'</span>, solver<span class="op">=</span><span class="st">'liblinear'</span>,penalty<span class="op">=</span><span class="st">'l1'</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="cf">with</span> plt.rc_context({<span class="st">"figure.figsize"</span>: (<span class="dv">8</span>, <span class="dv">12</span>), <span class="st">"figure.dpi"</span>: (<span class="dv">300</span>)}):</a>
<a class="sourceLine" id="cb13-3" data-line-number="3">    sc.pl.rank_genes_groups(adata_niches, n_genes<span class="op">=</span><span class="dv">5</span>, sharey<span class="op">=</span><span class="va">False</span>,show<span class="op">=</span><span class="va">True</span>,fontsize<span class="op">=</span><span class="dv">20</span>)</a></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-11-3.png" width="3072"></p>
<p>We can also check the interaction levels on violin plot for certain markers.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb14-1" data-line-number="1">marker_LRs <span class="op">=</span> np.array([g <span class="cf">for</span> g <span class="kw">in</span> adata_niches.uns[<span class="st">'rank_genes_groups'</span>][<span class="st">'names'</span>][<span class="dv">0</span>]])</a></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co"># check specific LRs</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="cf">with</span> plt.rc_context({<span class="st">"figure.figsize"</span>: (<span class="dv">10</span>,<span class="dv">7</span>), <span class="st">"figure.dpi"</span>: (<span class="dv">300</span>),<span class="st">"font.size"</span>:(<span class="dv">8</span>)}):</a>
<a class="sourceLine" id="cb15-3" data-line-number="3">    sc.pl.violin(adata_niches, marker_LRs[<span class="dv">0</span>], groupby<span class="op">=</span><span class="st">'VectorType'</span>,rotation<span class="op">=</span><span class="dv">20</span>)</a></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-13-5.png" width="1270"></p>
<p>We can also visualize the interaction levels on top of UMAP.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="cf">with</span> plt.rc_context({<span class="st">"figure.figsize"</span>: (<span class="dv">8</span>, <span class="dv">8</span>), <span class="st">"figure.dpi"</span>: (<span class="dv">300</span>)}):</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">    sc.pl.umap(adata_niches, color<span class="op">=</span>marker_LRs[:<span class="dv">2</span>],ncols<span class="op">=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-14-7.png" width="1710"></p>
<p>We can also visualize the interaction levels for the top markers of different groups using dot plot.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="cf">with</span> plt.rc_context({<span class="st">"figure.dpi"</span>: (<span class="dv">300</span>),<span class="st">"font.size"</span>:(<span class="dv">10</span>)}):</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">    sc.pl.dotplot(adata_niches, marker_LRs, groupby<span class="op">=</span><span class="st">'VectorType'</span>,figsize<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">10</span>))</a></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-15-9.png" width="1920"></p>
</div>
</div>
<div id="interporability-with-monocle3-celldataset" class="section level1">
<h1 class="hasAnchor">
<a href="#interporability-with-monocle3-celldataset" class="anchor"></a>Interporability with Monocle3 (CellDataSet)</h1>
<p>Here we will continue to use this pbmc3k dataset, but we will switch to Monocle3 (CellDataSet). This will be a simple demonstration of how to convert the output from NICHES to CellDataSet and use certain functions from Monocle3 to perform downstream analysis.</p>
<div id="convert-the-output-from-niches-to-celldataset" class="section level2">
<h2 class="hasAnchor">
<a href="#convert-the-output-from-niches-to-celldataset" class="anchor"></a>Convert the output from NICHES to CellDataSet</h2>
<p>Here we use the same output from NICHES from previous run and convert it to CellDataSet.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">monocle3</span><span class="op">)</span>
<span class="va">lr_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">lr_names</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">lr_metadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">lr_names</span>
<span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/new_cell_data_set.html">new_cell_data_set</a></span><span class="op">(</span><span class="va">niches_list</span><span class="op">$</span><span class="va">CellToCell</span><span class="op">$</span><span class="va">CellToCellMatrix</span>,
                         cell_metadata <span class="op">=</span> <span class="va">niches_list</span><span class="op">$</span><span class="va">CellToCell</span><span class="op">$</span><span class="va">metadata</span>,
                         gene_metadata <span class="op">=</span> <span class="va">lr_metadata</span><span class="op">)</span></code></pre></div>
</div>
<div id="data-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-processing" class="anchor"></a>Data Processing</h2>
<p>Then we will process the data using Monocle3 and run PCA and UMAP on it.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># use all genes for pca</span>
<span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html">preprocess_cds</a></span><span class="op">(</span><span class="va">cds</span>, num_dim <span class="op">=</span> <span class="fl">30</span>,
                      method <span class="op">=</span> <span class="st">"PCA"</span>,norm_method <span class="op">=</span> <span class="st">"none"</span>,scaling <span class="op">=</span> <span class="cn">TRUE</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html">reduce_dimension</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></code></pre></div>
</div>
<div id="data-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#data-visualization" class="anchor"></a>Data Visualization</h2>
<p>We can visualize the data on UMAP (Cells are colored by VectorTypes).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html">plot_cells</a></span><span class="op">(</span><span class="va">cds</span>, color_cells_by<span class="op">=</span><span class="st">"VectorType"</span><span class="op">)</span></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
</div>
<div id="interporability-with-scater-singlecellexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#interporability-with-scater-singlecellexperiment" class="anchor"></a>Interporability with Scater (SingleCellExperiment)</h1>
<p>Here we will continue to use this pbmc3k dataset, but we will switch to Scater (SingleCellExperiment) and utilize certain functions to perform downstream analysis.</p>
<div id="convert-the-output-from-niches-to-singlecellexperiment" class="section level2">
<h2 class="hasAnchor">
<a href="#convert-the-output-from-niches-to-singlecellexperiment" class="anchor"></a>Convert the output from NICHES to SingleCellExperiment</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="op">(</span>
    assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">niches_list</span><span class="op">$</span><span class="va">CellToCell</span><span class="op">$</span><span class="va">CellToCellMatrix</span><span class="op">)</span>, 
    colData <span class="op">=</span> <span class="va">niches_list</span><span class="op">$</span><span class="va">CellToCell</span><span class="op">$</span><span class="va">metadata</span>,
    rowData <span class="op">=</span> <span class="va">lr_metadata</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="runpca-on-the-ouput" class="section level2">
<h2 class="hasAnchor">
<a href="#runpca-on-the-ouput" class="anchor"></a>RunPCA on the ouput</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">sce</span>,exprs_values<span class="op">=</span><span class="st">"data"</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualize-ligand-receptor-interactions-using-violin-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-ligand-receptor-interactions-using-violin-plot" class="anchor"></a>Visualize Ligand-receptor interactions using violin plot</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html">plotExpression</a></span><span class="op">(</span><span class="va">sce</span>,features <span class="op">=</span> <span class="st">"GNAI2—C5AR1"</span>,exprs_values <span class="op">=</span> <span class="st">"data"</span>,x <span class="op">=</span> <span class="st">"VectorType"</span>,colour_by<span class="op">=</span><span class="st">"VectorType"</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">60</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div id="visualize-the-output-on-umap" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-the-output-on-umap" class="anchor"></a>Visualize the output on UMAP</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>,exprs_values<span class="op">=</span><span class="st">"data"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"VectorType"</span><span class="op">)</span></code></pre></div>
<p><img src="08%20Interoperability%20-%20NICHES_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Micha Sam Brickman Raredon, Junchen Yang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
