<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>RunNICHES — RunNICHES • NICHES</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="RunNICHES — RunNICHES" />
<meta property="og:description" content="Performs user-selected NICHES transformations on a Seurat object. 
By default, references RunCellToCell, RunCellToSystem, and RunSystemToCell functions. 
If positional data is provided, similar analyses can be performed which are limited exclusively to cells that directly neighbor each other.
Output is a set of specialized NICHES objects which allow fine analysis of cell-cell interactions." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">NICHES</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/01 NICHES Spatial.html">Estimating Microenvironment from Spatial Data</a>
    </li>
    <li>
      <a href="../articles/02 NICHES Single.html">Converting Single-Cell Atlases to Single-Cell Signaling Atlases</a>
    </li>
    <li>
      <a href="../articles/03 Rat Alveolus.html">Exploring Cellular Relationships in the Rat Alveolus</a>
    </li>
    <li>
      <a href="../articles/04 NICHES Differential By Vectortype.html">Using NICHES to Analyze Cell-Cell Signaling Across Conditions</a>
    </li>
    <li>
      <a href="../articles/05 Pseudotime.html">Pseudtotemporal Ordering with NICHES</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>RunNICHES</h1>
    
    <div class="hidden name"><code>RunNICHES.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs user-selected NICHES transformations on a Seurat object. 
By default, references RunCellToCell, RunCellToSystem, and RunSystemToCell functions. 
If positional data is provided, similar analyses can be performed which are limited exclusively to cells that directly neighbor each other.
Output is a set of specialized NICHES objects which allow fine analysis of cell-cell interactions.</p>
    </div>

    <pre class="usage"><span class='fu'>RunNICHES</span><span class='op'>(</span>
  <span class='va'>object</span>,
  assay <span class='op'>=</span> <span class='st'>"RNA"</span>,
  LR.database <span class='op'>=</span> <span class='st'>"fantom5"</span>,
  <span class='va'>species</span>,
  min.cells.per.ident <span class='op'>=</span> <span class='cn'>NULL</span>,
  min.cells.per.gene <span class='op'>=</span> <span class='cn'>NULL</span>,
  meta.data.to.map <span class='op'>=</span> <span class='cn'>NULL</span>,
  position.x <span class='op'>=</span> <span class='cn'>NULL</span>,
  position.y <span class='op'>=</span> <span class='cn'>NULL</span>,
  custom_LR_database <span class='op'>=</span> <span class='cn'>NULL</span>,
  k <span class='op'>=</span> <span class='fl'>4</span>,
  rad.set <span class='op'>=</span> <span class='cn'>NULL</span>,
  blend <span class='op'>=</span> <span class='st'>"mean"</span>,
  CellToCell <span class='op'>=</span> <span class='cn'>T</span>,
  CellToSystem <span class='op'>=</span> <span class='cn'>F</span>,
  SystemToCell <span class='op'>=</span> <span class='cn'>F</span>,
  CellToCellSpatial <span class='op'>=</span> <span class='cn'>F</span>,
  CellToNeighborhood <span class='op'>=</span> <span class='cn'>F</span>,
  NeighborhoodToCell <span class='op'>=</span> <span class='cn'>F</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Seurat object. The active identity will be used to define populations for connectomic sampling and crossings.</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p>string. Default: "RNA". The assay to run the NICHES transformation on.</p></td>
    </tr>
    <tr>
      <th>LR.database</th>
      <td><p>string. Default: "fantom5". Currently accepts "fantom5","omnipath", or "custom"</p></td>
    </tr>
    <tr>
      <th>species</th>
      <td><p>string. The species of the object that is being processed. Only required when LR.database = 'fantom5' with species being 'human','mouse','rat', or 'pig', or LR.database = 'omnipath' with species being 'human','mouse', or 'rat'</p></td>
    </tr>
    <tr>
      <th>min.cells.per.ident</th>
      <td><p>integer. Default: NULL. A limit on how small (how many cells) a single population can be to participate in connectomic crossings.</p></td>
    </tr>
    <tr>
      <th>min.cells.per.gene</th>
      <td><p>integer. Default: NULL. Limits analysis to interactions involving genes expressed above minimum threshold number of cells in the system.</p></td>
    </tr>
    <tr>
      <th>meta.data.to.map</th>
      <td><p>character vector. Optional. Default: NULL. A vector of metadata names present in the original object which will be carried to the NICHES objects</p></td>
    </tr>
    <tr>
      <th>position.x</th>
      <td><p>string. Optional. Default: NULL. The name of the meta.data column specifying location on the spatial x-axis. Only required for spatial omics data.</p></td>
    </tr>
    <tr>
      <th>position.y</th>
      <td><p>string. Optional. Default: NULL. The name of the meta.data column specifying location on the spatial y-axis. Only required for spatial omics data.</p></td>
    </tr>
    <tr>
      <th>custom_LR_database</th>
      <td><p>data.frame. Optional. Default: NULL. Only required when LR.database = "custom". Each row is a ligand-receptor mechanism where the first column corresponds to the source genes that express the ligands subunits (separated by '_') and the second column corresponds to the receptor genes that express the receptor subunits (separated by '_').</p></td>
    </tr>
    <tr>
      <th>k</th>
      <td><p>integer. Optional. Default: 4. Number of neighbors in a knn graph. Used to compute a mutual nearest neighbor graph based on the spatial coordinates of the spatial transcriptomic datasets.</p></td>
    </tr>
    <tr>
      <th>rad.set</th>
      <td><p>numeric. Optional. Default: NULL. The radius threshold to define neighbors based on the spatial coordinates of the spatial transcriptomic datasets. Ignored when 'k' is provided.</p></td>
    </tr>
    <tr>
      <th>blend</th>
      <td><p>string. Default: "mean". Choice of linear operator to combine edges in single-cell niche investigations. Defaults to "mean", also accepts "sum".</p></td>
    </tr>
    <tr>
      <th>CellToCell</th>
      <td><p>logical. Default: TRUE. Whether to analyze cell-cell interactions without considering spatial coordinates.</p></td>
    </tr>
    <tr>
      <th>CellToSystem</th>
      <td><p>logical. Default: FALSE. Whether to analyze summed signaling output to total system coming from each cell. Does not consider Euclidean coordinates.</p></td>
    </tr>
    <tr>
      <th>SystemToCell</th>
      <td><p>logical. Default: FALSE. Whether to analyze summed signaling input from total system landing on each cell (cellular microenvironment/niche). Does not consider Euclidean coordinates.</p></td>
    </tr>
    <tr>
      <th>CellToCellSpatial</th>
      <td><p>logical. Default: FALSE. Whether to analyze cell-cell interactions between Euclidean neighbors. Only applicable in spatial datasets.</p></td>
    </tr>
    <tr>
      <th>CellToNeighborhood</th>
      <td><p>logical. Default: FALSE. Whether to analyze summed signaling output to Euclidean neighbors. Only applicable in spatial datasets.</p></td>
    </tr>
    <tr>
      <th>NeighborhoodToCell</th>
      <td><p>logical. Default: FALSE. Whether to analyze summed signaling input from Euclidean neighbors (cellular microenvironment/niche). Only applicable in spatial datasets.</p></td>
    </tr>
    </table>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Micha Sam Brickman Raredon, Junchen Yang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


